### DISCUSSION TITLE
C# source generators

### DISCUSSION THREAD

**Discussion by celeronpm** [nxrighthere/UnrealCLR#171] [Ideas]
REDACTED

https://www.change.org/p/tell-epic-you-want-to-see-code-orchestra-in-unreal-engine

**Comment by ScottKane** [comment]
That is absolutely brilliant For me, trying to move from Unity, I've encountered the same problems. Just wishing I could have a derivable in-engine class to work with to tie together blueprints and C# through inheritance rather than awkward back and forward static method invocations (at least I assume the generated C++ classes play nice). This for me would be a game changer.

**Comment by nxrighthere** [comment]
Nice work! A few questions if you don't mind:

- Do you reload a module with generated code dynamically while the editor is running?
- If yes, what approach do you use to track changes and trigger compilation of the generated code?
- Is hot-reload works consistently in your case when you consecutively add/remove properties after initial code generation?
- If generated code is used in the game module just like a regular C++ code, how do you handle cases when it's impossible to reload it (for example, with a blueprint function library)?

**Reply by celeronpm** [reply]
- Yes, our Visual Studio build simply pushes the new built managed DLL's out to the UE project directory and also pushes the C++ class changes to the codebase (overwrites if existing). We use a task runner (can also be done as a build step for the c# library) for this and automatically call the Unreal Build batch file as part of the process to rebuild the unreal code. This gets us from a C# class change to a usable ready-to-go state in a few seconds in a lightweight proof of concept project.
- Yes, we haven't experienced any issues with hot-reload since we are rebuilding the code. We have also made changes to how UnrealCLR manages the dynamically loaded managed libraries because we have code that needs to live for the entire lifecycle of the executable, not just per level (EG. a custom MMO networking library we built on top of DarkRift2 networking that needs to maintain sockets to our MMO back end for the entire life of the game).
- We are avoiding blueprint development so haven't had this issue come up / have had no reason to look into it. I am sure we can dig in further. Coming from Unity, we are not used to visual scripting since C# makes it so quick and easy to get complex logic in, it feels unnatural (even though I understand why some folks prefer it over C++ / either use a language that is half a century old or drag and drog logic and deal with the complexities of function libraries instead - I guess that is the reason we are exploring UnrealCLR to begin with!).

**Reply by ScottKane** [reply]
I think the use case for most people with blueprints will just be to integrate marketplace assets and third party plugins which I think will be covered as you can just fire the events in C# for whatever the blueprint needs. Being able to have the blueprint derive from a generated C++ class would improve the current workflow

**Reply by celeronpm** [reply]
I can definitely see that. Worst case, if we can't get dynamic reload for it, a UE restart would reload and expose the endpoints. 

Since Unreal is open source and I am in the honeymoon phase, I am at the peak of inflated expectations and believe we can easily solve this. Having an open source engine changes the game and feels too good to be true.

We ran into so many dead ends with the closed Unity implementation we're about to dump 2 years of work and just restart with UE. We built our own Nanite like quantized mesh rendering system and just couldn't get the SRP to bow to our will fully due to closed source bs (we got it rendering meshes with tens of billions of triangles in HDRP but editor integration was not ideal). Also the DXR implementation in Unity is beyond a joke, and so is their custom mono runtime. 

I'd rather invest time in pursuing our code gen approach and have the ability to leverage the source of the engine if needed vs hitting wall after wall.

**Reply by ScottKane** [reply]
I completely agree, I got sick of everything being so ridged with the old full framework. No nuget packages built in easily. A poor component system that restricts creativity by removing construction in favour of an approach that makes it easier for scripters. I implemented DI with custom constructors and set up an in memory sqllite db before getting frustrated and moving to UE. All before source generators, not that it matters because you couldn't easily use C#9 and certainly not .NET 5. Really hope this gets integrated or to see a fork in the future. Would be a good learning experience for working with generators between C# and C++

**Reply by MostHated** [reply]
I have been (and still am) a huge advocate of this (addon/package/system?) since day one, and these new enhancements you guys are talking about sound fantastic, but one of the biggest "hangups" for me taking the plunge, so to speak, (aside from Epic seeming to 'tolerate' linux users at best when it comes to the editor and lack of launcher) is that Unity's Job system has been an absolute game-changer for me. 

Being able to essentially lob a bunch of work at all cores with a relatively minimal amount of effort, while having compile and runtime safety measures and leak detection has major staying power. I might feel differently if I could try out nanite and lumen on Linux, but even with as awesome as these things sound (especially performance improvements over the mono implementations), losing the ability to parallelize things so effortlessly, I just find that really hard to give up. Simply knowing that it isn't an option has kept me from pursuing any more than just "playing around" with things here and there as opposed to giving it an actual try with a project.

**Reply by celeronpm** [reply]
We already had a multi-threading/job system in place because we started our project before the job system was available in Unity. Not only do we have something that is a little bit more flexible, we have rebuilt the couroutine concept for Unreal at this point. 

It's all integrated into a message bus system that supports fan-out via parallelism and main thread re-join. TLDR; not only can you rebuild the job system, but you can do it much better for UE. 

And that's the beauty of this, you can do it so much better than what the closed Unity implementation offers. Only if they open sourced their trash, they could benefit from so many folks.

**Reply by ScottKane** [reply]
I wrote a tiny event bus using reactive extensions, they can be streamed and wouldn't take much to pull it up to run asynchronously on multiple threads

**Reply by celeronpm** [reply]
The multi threaded job system extends to the task system to achieve this, looks like Unity huh?

public static IEnumerator SomeTask() {
  yield return Tasks.AwaitNextFrame();
  yield return Tasks.WaitSeconds(...);
}

**Reply by MostHated** [reply]
Interesting... what are the chances that you might be considering releasing it at some point? Something like that, mixed with all of the above mentions really ticks a lot of the 'apprehension boxes' as far as I am concerned. The only other concern I really had was more about just the types of things I enjoy doing, and probably isn't even a topic for this particular conversation, but for every hour I spend working on an actual game, I spend 10-20 hours building editor tools. 

One thing I was never quite sure of was if dotnet could be leveraged for actual editor interaction and manipulation or vice versa to be able to create tools full in C# in a way similar to Unity EditorWindow, Inspector tools, hooks into pre/post processing events, asset database, etc. I am not sure if any of that has changed, be it for better or worse, with UE5 either, but just the fact that it was mentioned that the dotnet side of things was able to be manipulated via UE editor windows made me somewhat excited to see.

**Reply by celeronpm** [reply]
Check out our project, www.CrushDepth.com

We had had to do internal 64bit vectors and translate via floating origin to 32bit vectors like Kerbal Space Program. With Unreal supporting 64 bit double precision transforms now (or really soon) in the entire stack (including physics), it's a game changer for our world scale MMO.

Unigine was the only other C# engine that supported 64 bit, but let's not go there... 😄

**Reply by celeronpm** [reply]
> The only other concern I really had was more about just the types of things I enjoy doing, and probably isn't even a topic for this particular conversation, but for every hour I spend working on an actual game, I spend 10-20 hours building editor tools.
> 
> One thing I was never quite sure of was if dotnet could be leveraged for actual editor interaction and manipulation or vice versa to be able to create tools full in C# in a way similar to Unity EditorWindow, Inspector tools, hooks into pre/post processing events, asset database, etc. I am not sure if any of that has changed, be it for better or worse, with UE5 either, but just the fact that it was mentioned that the dotnet side of things was able to be manipulated via UE editor windows made me somewhat excited to see.

We've got editor extensions in our backlog of items to investigate. A simple webrenderer component with a typescript/angular UI talking to your lib which exposes a rest API would make this easily achievable if all else fails. But given UE has classes and transaction scopes already, I'm sure we can make it happen with our codegen approach. Just haven't needed it yet since we're not at the point of custom forms in our proof of concept development phase.

**Reply by MostHated** [reply]
Unigine seemed like a great option right out of the gate, but it just seemed like it was going to end up being a waste of time to put any sort of real time into it as an indie developer. Too many of the features that would have made it worth using were paywalled beyond what I was willing to spend to pursue a passion, even being as dedicated as I am.

**Reply by ScottKane** [reply]
I would recommend a blazor webassembly, keeps you in .NET and you can strip away all of the web side of things by using a HostBuilder where you start up your events system

**Reply by celeronpm** [reply]
I think we're all agreeing that we can make UE .Net friendly at the end of the day, so far it is easier than we anticipated.

**Reply by MostHated** [reply]
All of this really feels like a whole new door may be opening for what this project is capable of achieving, and actually has me quite excited simply knowing that a lot of the things I felt I would be missing out on might one day not only be present, but even better than I had hoped.

**Reply by ScottKane** [reply]
I'm really glad to see this happening, I have a basic understanding of C++ and hatred of mangled visual scripting. Using C# in Unreal is so nice

**Reply by celeronpm** [reply]
> Interesting... what are the chances that you might be considering releasing it at some point? 
> 

Not sure to be honest. Theoretically in a few months, once our initial port of all core frameworks is complete, we will have production quality frameworks ready to go. Once they are fully integrated and thoroughly tested in our alpha release that we are porting from Unity, we can think about making them available.

**Reply by celeronpm** [reply]
We have also applied for an Unreal Megagrant like this project. If we also get it, I think we could release and support our frameworks as plug and play extensions to UnrealCLR.

**Reply by ScottKane** [reply]
I hope you get it. If this all gets off the ground I'm in for a great experience in Unreal

**Reply by nxrighthere** [reply]
I implemented a similar approach a few months ago, but module reload didn't work consistently in a few test cases. So I started to work on an in-memory generation of classes and properties at runtime similar to Python implementation from Epic, but lack of control over the engine's code and .NET as a middleware plugin led to other limitations.

In general, I like the idea of code generation, but it should meet at least the following requirements:

1. Code generation should be independent of any IDE, and work for all supported platforms, it means tracking of changes and triggering of code compilation should be implemented in the plugin's runtime with [`DirectoryWatcher`](https://docs.unrealengine.com/4.26/en-US/API/Developer/DirectoryWatcher/) and [`FModuleManager`](https://docs.unrealengine.com/4.26/en-US/API/Runtime/Core/Modules/FModuleManager/) API or similar. 
2. Module reload for compiled code should work transparently and consistently for the end-user with data types that we currently support in the [`Object`](https://github.com/nxrighthere/UnrealCLR/blob/30c29682c5b34d3e8bdb365ff9ef9e161d78c727/Source/Native/Source/UnrealCLR/Public/UnrealCLRFramework.h#L338) API.
3. Unsafe code is acceptable in the framework but should be avoided in user code, only managed references or safe blittable data types.

**Reply by celeronpm** [reply]
1. We are testing CodeGen with roslyn as a separate project that can be executed in the pipeline without the IDE integration requirements; this would have the benefit that the user can inspect the generated code and see exactly what is being created.
2. Again, I don't see issues there
3. Codegen can use a compiler flag to switch between managed references vs blazing-fast pointers. 

All of the above are doable.

**Reply by nxrighthere** [reply]
Sounds interesting to me. If you provide your implementation for testing, and if it works as expected, I'll merge it. Thanks!

**Reply by ScottKane** [reply]
Am I right in thinking you said you have the slate API working? Would be good to see this for networking too so we can wrap methods in Client, Server and Multicast attributes

**Reply by celeronpm** [reply]
Slate, not yet, our networking solution is built off of darkrift2. 

Because we built our MMO layer using darkrift2, we got the networking working in UE 5 really quickly. We spin up a custom dispatcher based client and simply use our custom networking layer to spawn blueprints, synchronize positions, call RPC methods, etc...

I would not be comfortable using the built in networking system due to the extreme use cases we have, so it's unlikely we'll spend any time on it.

**Reply by ScottKane** [reply]
Sure makes sense, the generators could be updated if merged into UnrealCLR to handle all of the C++ macros if they don't already. From a usability standpoint it would be nice to do things like you would in C++ but from C#.

For most users I would imagine the engines networking API would do the job but obviously an mmo networking stack is a different ball game

**Reply by celeronpm** [reply]
An update, we've moved our code gen from the c# source generators to a custom console project using the Roslyn apis. This allows us to do the codegen in any IDE and allows us to build a cleaner build pipeline. All we do is call the exe with some args and it loads the soliton, walks the code, and extends the c# classes where appropriate while generating the c++ proxies. It also handles cleanup, renaming or classes, etc via a local cache (which also speeds up rebuilds).

Since source generators are built on Roslyn, we used about 98% of the code, just added extra bootstrapping harnesses to load the compile context from a workspace instead of the source generator injection.

**Reply by celeronpm** [reply]
So at this point we have a player that can walk around, look at a usable object, interact with it, spawn networked objects, control Network object position and movement, request Network object ownership, and some of the other basics you would expect to have in an MMO.

All in C# (ported most of our unity logic, the refactoring to fit the actor/scene/primitive concepts has been pretty straightforward).

**Reply by ScottKane** [reply]
That's really encouraging, how are you handling ticking? Do you have your own event system handling ticks or does your C++ generated code point to a pointer in C# to run a tick method?

**Reply by Friedrich-S** [reply]
Ticking is handled entirely in C# which makes it fast and gives great flexibility when it comes to custom checks.

**Reply by celeronpm** [reply]
As @FriedrichLP mentioned, we chose to have a global manager on our end which is responsible for the lifecycle of the c# object, including ticking (so we can achieve a Unity like tick priority order based on type and any other logic we need).

However, we would be able to create a c++ tick bridge into c# code if we wanted to, we just think that isolating the code and putting maximum control in managed land makes the most sense for our use case.

**Reply by celeronpm** [reply]
We will keep this thread going so we can show what is possible with c# in Unreal and post some links to devlogs eventually if that's ok with @nxrighthere otherwise we can move this over to our own forums.

**Reply by MostHated** [reply]
It's all pretty relevant, and this is the discussion section, so I would not expect it to be an issue. I definitely like getting notified when there are new posts to keep up with the conversation and happenings.

**Reply by nxrighthere** [reply]
@celeronpm Feel free to discuss here anything that is related to the plugin. Also, if you have some gameplay demo baked by the plugin, we have a showcase category here.

**Reply by ScottKane** [reply]
@celeronpm any idea when we could expect a merge of your generation features?

**Reply by celeronpm** [reply]
@ScottKane please DM on Discord -  `CrushDepth#0001`

**Reply by MostHated** [reply]
I would definitely be interested in hearing as well.

**Reply by celeronpm** [reply]
> I would definitely be interested in hearing as well.

DM me on discord please.

**Reply by darkguy2008** [reply]
@celeronpm any way we can get updates to this without getting into discord and DMing you? I feel it's too bloated. The discussion was great and I'd like to see more news coming in from here. Thanks!

**Reply by celeronpm** [reply]
@darkguy2008 yes, I can ping you here if/when we have a path forward for distribution. Given the complexity of the situation, there are no guarantees we can proceed as we had initially planned, unfortunately. I am extremely grateful to @pixeltris for bringing clause G of the EULA to my attention and hope we can find a resolution.

**Comment by celeronpm** [comment]
Hi all, we've decided to move forward with a custom .NET solution which fits our needs more closely and addresses unique requirements we have for our large scale high fidelity project. Please close this discussion if no longer needed and/or delete this comment if not appropriate.

Our solution is called code orchestra and can be found at https://code-orchestra.com

**Reply by darkguy2008** [reply]
Hi! So this means you're not proceeding with the code merge of the automatic code generation in the UnrealCLR project? :(

Also, how is code-orchestra different from UnrealCLR? Is it based off it? Or is it something else altogether? Will it be free no-strings-attached like UnrealCLR is?

**Reply by celeronpm** [reply]
TLDR; we ended up needing something faster with unsafe memory access and bi-directional interlanguage capabilities with more UE API coverage. We built our own framework based off of heavy code-gen. Completely different approach then UnrealCLR.

More info on the site https://code-orchestra.com and on our Discord.

**Reply by pixeltris** [reply]
@darkguy2008 the Unreal Engine EULA requires all programming language integrations to be free and open source.

`g. Distribution of Non-C++ Programming Language Integration - You may Distribute an integration of a programming language other than C++ for the Licensed Technology, but if you do, the integration must be Distributed free of charge to all Engine Licensees, must be available in source code form (including, but not limited to, any compiler, linker, toolchain, and runtime), and must permit Distribution free of charge, on all platforms, in any Product.`

https://www.unrealengine.com/en-US/eula/publishing

**Reply by celeronpm** [reply]
Thanks for the EULA link @pixeltris

We've got three routes we are pursuing here:

1. A custom license request with Epic
2. Megagrant application
3. Marketplace Distribution

We believe that the [Epic Marketplace Distribution EULA](https://www.unrealengine.com/en-US/marketplace-distribution-agreement) extends the [Unreal Engine EULA](https://www.unrealengine.com/en-US/eula/publishing) with additional clauses which make monetizing such a plug-in possible.

Examples:

- https://www.unrealengine.com/marketplace/en-US/product/jbi-java-blueprint-interface  
- https://www.unrealengine.com/marketplace/en-US/product/luamachine
- https://www.unrealengine.com/marketplace/en-US/product/python-editor

All 3 of the aforementioned monetized plugins are language extensions and would, outside of the marketplace distribution EULA, have to comply with the UE EULA and would thus have to be open-sourced. We believe the clause you called out is there to ensure all such tooling is sold exclusively through the Epic Marketplace and not monetized or licensed externally. We hope that one of the three avenues we are pursuing land so we can move forward with distribution. Until that point in time, we are collecting sign-ups from interested devs and discussing our approach, but are not sharing any tooling.

If we are unable to monetize our efforts to support the dev and maintenance costs of the tooling, we will most likely bin the efforts of bringing the toolchain to a broader audience for now (to answer @darkguy2008 question fully). While it is still too early to tell, existing plugins show that there is a path forward.

**Reply by darkguy2008** [reply]
I see, thank you both for your answers! I appreciate the detailed explanations.

I was very happy when I read @pixeltris 's comment but I guess I got excited too early. I've been longing for the day when I can fully use UE with C# to get rid of Unity... Hopefully those days will be over with Code Orchestra's release. Don't get me wrong, UnrealCLR is pretty amazing so far, although with some quirks thanks to .NET's limitations. @celeronpm your approach is different so it'd be good to check it out too when it's released.

Well, developers gotta eat too, yes. I'll hope for the best, so Epic can fund you guys too or something so we can enjoy it the same way we already enjoy UnrealCLR :). Thanks and best of luck! Looking forward to it :D

**Reply by celeronpm** [reply]
Thanks @darkguy2008. We just stumbled across https://forums.unrealengine.com/t/i-want-feedback-from-epic-about-mono-for-unreal-engine/17343/17

It is not looking very promising for a sustainable path forward for a properly supported and maintained release due to the roadblock of all monetization options. That answer explains why projects like MonoUE and USharp eventually get abandoned by the contributors.

@pixeltris thank you again for bringing this to my attention! We've got quite a bit to consider as our whole project relies on Code Orchestra and under the clause, we would be forced to make it free and open source if we want to release the title on UE.

May I ask, is this why you never expanded USharp?

**Reply by pixeltris** [reply]
I'm sure you're right about the marketplace for monetization and it sounds like a good idea. Maybe you could clarify with Epic to be sure. It sounds like it should be an amendment to the EULA if so.

The EULA explicitly mentions *distribution* of programming language integrations. So you're fine if you're integrating a programming language into a closed source game.

That EULA change happened while MonoUE was in development by Xamarin so they ceased development and it became a hobby project of the devs. Additionally the lack of action by Epic to accept PRs https://github.com/mono-ue/UnrealEngine/issues/208#issuecomment-776903851 is an issue for making it a standalone plugin which made development not so fun and reduced potential contributors.

USharp has a huge amount of issues. For example:
- Hotreload. C++ hotreload is (was?) broken. Simple properties are "generally" fine. But complex types, especially components aren't properly handled by the reinstancing step of hotreload. i.e. https://github.com/pixeltris/USharp/issues/39
- No native storage for holding a reference to the C# object for UObject types. This means forking the engine and inserting a pointer / GCHandle to the C# object in UObject/UObjectBase. USharp currently uses a [slow array or dictionary lookup](https://github.com/pixeltris/USharp/blob/8f0205cd8628b84a66326947d93588a8c8dcbd24/Managed/UnrealEngine.Runtime/UnrealEngine.Runtime/Internal/GCHelper.cs#L98-L123). UnrealCLR handles this by constructing a new object every time a UE API is called which has its own issues https://github.com/nxrighthere/UnrealCLR/issues/61#issuecomment-721366024. Another way to handle this problem is creating a "C#" equivalent of every C++ type (e.g. AManagedActor, UManagedSceneComponent, etc) where you can hold onto the GCHandle. I think this is what https://github.com/mrkriv/UnrealDotNet does. But if you try to access a native C++ object which isn't a "UManagedXXX" type you'd still have the same issue.
- Marshaling of arrays and structs. Currently with USharp if you expose a C# function which takes an array, that entire array is copied from C++ to C#, then back from C# to C++. There are ways to handle this by using custom types such as TArray, TMap. But then if you were to call that function from C# you're then constructing a TArray/TMap so you'd need those types to also have a backing C# List/Dictionary and on every accessory e.g. `[]`, `ContainsKey`, etc, you'd need to check if you should be doing native memory lookups or using the List/Dictionary. Structs also have the same issue. Idealy if you have a large struct you don't want to be copying the entire thing from/to a C# compatible type. Again, you could use pointers and use C# properties to read native memory, but in this case where you are calling a C# function from C# which takes a struct you'd then need to construct and manage the memory for the struct manually.
- The code generator is a mess. The thing which collects info about C# types and constructs a UClass type mapping info is a mess. The UClass builder is a mess. `AssemblyRewriter` is a mess, slows down compilation times, and breaks edit-and-continue debugging.
- Too many things are currently generated both for accessing C++ exposed things and exposing C# things. Some improvements could be made but ultimately offsets of memory need to be determined at runtime and cached *somewhere* for C# to access.
- Some things just don't map that well to C#. For example FText is reference counted which is a really minor thing and works great from C++ land. But from C# it means using IDisposable and being careful. There are always finalizers, but the C# GC doesn't know how much memory FText represents. Delegates were a mess to implement and don't really map that well to their C++ counterparts. Interfaces are a mess. No idea how I'd handle exposing Slate which are all referenced counted types, maybe UMG would be easier.
- There are some things you really want to access / override from C# but simply aren't exposed. Looking at MonoUE's PRs I decided to create [VTableHacks](https://github.com/pixeltris/USharp/blob/8f0205cd8628b84a66326947d93588a8c8dcbd24/Managed/UnrealEngine.Runtime/UnrealEngine.Runtime/Internal/VTableHacks.cs) which is as bad as it sounds.

Ultimately I think it was the hotreload issues and tech debt which caused me to stop work on USharp. The hotreload issue was essentially impossible for me to fix without maintaining an engine fork. If C# hotreload is to be perfect, the C++ hotreload needs to be perfect. And if hotreload is a no-go then it really destroys the nice development iteration when using C#.

UnrealCLR avoids many issues by simply not having such a deep integration with the engine. But it needs to (at some point) address this to create a better workflow. At minimum it needs #33 and idealy also #61.

When I archived the USharp repo I made this [gist](https://gist.github.com/pixeltris/1c50f2333bba9f3f3f3b2e30e99cac7a) talking about this, but unfortunately haven't got around to touching UnrealCLR. I'm still not sure what the ideal integration would be, but I think some kind of code gen so that C# can access Blueprint properties / functions, as well as expose C# functions to Blueprint. Possibly "partial" generated classes where you write your own functions from C# for a given Blueprint. This is essentially the approach this project took https://github.com/mihe/tsu

I might return to USharp at some point depending on what happens to the UE C# space and Epic's [Verse](https://reddit.com/r/unrealengine/comments/kf8z27/epic_showed_off_their_new_unreal_verse_scripting/). I like how USharp maps nearly 1:1 to UE C++ and I think that has some value, especially for people who want to try UE without using C++ or BP.

Your approach of code gen sounds pretty good. It sounds like iteration times would be reduced a bit for structural changes to types due to the C++ code gen / compilation times? But otherwise, yea it sounds like a pretty solid approach.

EDIT: To actually answer your question: I didn't stop USharp due to the EULA. MonoUE did mostly stop due to the EULA change. I considered the marketplace for USharp but due to the mentioned issues I never made it happened as I never got it to the point where I'd recommend its use.

**Reply by nxrighthere** [reply]
@pixeltris Such a good overview. Thanks for that.

Yes, you are totally right, and that's why I don't merge my experimental code related to https://github.com/nxrighthere/UnrealCLR/issues/61 and https://github.com/nxrighthere/UnrealCLR/issues/33 into the master in the current state. I'll end up dealing with the same problems.

I had a talk with Epic about issues with .NET integration and the scripting language they are inventing. They encountered a lot of similar challenges. So I asked them to solve pain points with the engine in interop friendly way for other programming languages, so we all could benefit from that.

I could solve some of the challenges by maintaining a custom fork of the engine, but I'm not going to do that. It's a lot of work.

.NET itself also has many limitations, but I'm a small middleware entity to convince Microsoft to work on those.

**Reply by celeronpm** [reply]
Thank you both @pixeltris and @nxrighthere for the extra clarity and color.  We have not run into any of the issues mentioned above. 

I think we benefitted from coming in completely ignorant to the challenges and approaches of existing solutions such as USharp, MonoUE, and UnrealCLR. I assume the reason is that we took a drastically different approach to the problem and unwittingly avoided the challenges encountered with the aforementioned projects through a novel approach. 

As is evident from this discussion, we have not been at this very long (a little over 2 months in total). After having spent a few days with UnrealCLR, it became clear that the framework would provide about 5% of the functionality we needed. We stepped back, looked at all of the features we wanted to achieve (C# integration, rust integration, interlanguage high-speed communication, shared data access to common memory to avoid marshaling (yes we keep memory offset pointers so we can access the memory directly with unsafe - it is how we achieve polyglot multi-language shared memory access), etc...) and started building a solution to accommodate the needs of our project.

We then used our proof of concept and quickly transplanted a ton of frameworks from our unity build into UE. Before we knew it, we had a UE build connecting to our MMO backend using DarkRift 2 transports and a ton of transplanted frameworks using Code Orchestra. We had clients synchronizing with our custom network sync system, had camera and item interactions working, had C# perform transform updates, etc...

Additionally, we started enhancing our add-on frameworks and started writing a clean ECS to offer natively for Unity devs coming to UE which have a proclivity for DOTS (specifically the benefits by writing scalable system-based routines instead of monobehaviors).  We have some incredible alpha testers signed up to run our solution through the gauntlet and help us ensure it scales well beyond our project, but at this point are blocked by clause G that we need to address.

I am hopeful that we can find a solution to this with Epic's support as this could be an incredible opportunity for the community.

**Reply by pixeltris** [reply]
@nxrighthere it's good that .NET has gotten interop improvements in recent years, but yea there's so much more that could be done which is a little frustrating.

@celeronpm it might be that the hotreload issues are now mostly solved which would be great. There used to be quite a few issues on their issue trackers which are no longer there. https://issues.unrealengine.com/issue/UE-52220 was a big one which looks to be fixed in 4.26.

Say you add a C# property to your class and press build, how long does it take in total reload into UE? Does it get worse when have a lot of classes? C++ compile times is one of the reasons I never really considered this approach.

Another reason I didn't consider that approach is because there isn't any technical reason why you couldn't do the same stuff at runtime (by construing UClass instances, rather than being compiled in via C++) and get the same performance in terms of accessing memory directly etc. Though there is a lot of complexity in doing simple things like overriding functions. It's a lot easier to directly invoke the C# function from C++ and avoid all the UFunction issues.

But yea, it's a really interesting approach and I'm interested to see how it develops! Hopefully you can get in contact with Epic and work something out.

**Reply by celeronpm** [reply]
"Say you add a C# property to your class and press build, how long does it take in total reload into UE? " For our project a few seconds. We've built intermediate caching mechanisms where possible. The c++ proxies are lightweight and the build is quick.

Code gen has many benefits over runtime approaches, I can't go further into the details at this moment as it is core to our IP.

**Reply by pixeltris** [reply]
A few seconds sounds great. If you could get say a ~3 second hotreload for a structural modification on something on the size of [ShooterGame](https://docs.unrealengine.com/4.27/en-US/Resources/SampleGames/ShooterGame/) or [UnrealTournament](https://github.com/EpicGames/UnrealTournament) that would be great and definitely the iteration cycle people would be looking for in a C# integration.

And yea I can imagine this type of code gen gives you a lot more control. I've definitely felt the limitations of the UObject system while working on USharp. For me it came down to compile times and having the ability to use C# without a C++ build system (i.e. BP + C#, or just C#). I guess I was wrong about the compilation times! I guess I should have at least tested...

If iteration times are good it sounds like a good choice for UnrealCLR too.

**Reply by celeronpm** [reply]
One thing I'd like to note also is that our codegen is bidirectional. Existing UE APIs are accessed via memory pointer offsets from C# proxies.

We do this at scale to ensure we can provide UE engine version specific supports including 3rd party plugin support (access 3rd party c++ class from c#, etc).

**Reply by MostHated** [reply]
It's a bit of a shame how much of a turn this thread took. Going from talk of merging awesome additions to the library, to "I can't go further into the details at this moment as it is core to our IP". I mean, I get it, but still.

**Reply by celeronpm** [reply]
@MostHated I definitely feel your frustration and thank you for the understanding. I would definitely PR the changes if they were changes to UnrealCLR. The only thing Code Orchestra and UnrealCLR have in common is that they provide C# functionality for UE devs. That's where the commonality ends and why 'merging' isn't actually possible. 

Code Orchestra is a project that was born out of necessity to solve a problem that still exists for our use case. It just so happens we first tried UnrealCLR and started sharing our ideas to solving these problems publicly here before we went full force on rethinking how the problem should be approached.

I am also completely open to removing this entire discussion from UnrealCLR to begin with; I am not comfortable taking up so much time and space away from the project on its own discussion thread. It is quite disrespectful from my perspective but at the same time, I feel compelled to respond to the discussion I started as long as @nxrighthere is OK with it. The moment he feels there is no longer value in this thread, I encourage him to delete it, archive it, or take any action he feels appropriate as the project owner.

At a minimum we've provided value by confirming approaches which were not fully explored by UnrealCLR or USharp have benefits. Both projects may benefit from the noise we've created here.
